Support for Explicit Polymorphism

Crazycolorz5 opened the pull request:
See #1030. Changed target branch after rebase as to not lose code in poly-adt-after relevant to possible implementation of Debrujin Indices.

disconcision committed:
merge in poly-adt

AlienKevin committed:
Fixed body_ctx of Forall and Rec in Statics

AlienKevin committed:
Allow Indet on the left of TypAp in Evaluator, similar to Ap

AlienKevin committed:
Fix mode of TypAp to Syn, not SynFun

AlienKevin committed:
Update is_fun and is_fun_var to recognize forall

Crazycolorz5 committed:
Implement semantics of type function application as type substition on terms.

Crazycolorz5 committed:
Add synthesis mode for type function and add necessary casting cases.
Add ground case for forall type (as Forall _ . ?).

Crazycolorz5 committed:
Add rule for evaluation of type application on cast.

Crazycolorz5 committed:
Some resolutions to conflicts from rebasing.

Crazycolorz5 committed:
Resolving issues in Statics and Elaborator.

Crazycolorz5 committed:
Edit constructors in frontend.

Crazycolorz5 committed:
Change dhexp substitution to use binding names rather than debrujin.

Crazycolorz5 committed:
Fix issue with looking up tvars in ctx.

Crazycolorz5 committed:
Fix error in alpha equiv

Crazycolorz5 committed:
Fix typechecking with mismatching type variable names.

Crazycolorz5 committed:
Implement capture avoiding substitution.

Crazycolorz5 commented:
To account for an analyzed type having a different binding than the expression being analyzed against, we perform alpha renaming via capture-avoiding binding. This can expose internal fresh type variable names to the user (see the screenshot below). This _might_ be able to be remedied by having the Forall constructor have an extra field that tracks the original binding name, and using the original name for all user-facing output (and possibly the same for Var).

![image](https://github.com/hazelgrove/hazel/assets/1497003/ac7d6092-075b-4923-b28a-5a3611111846)


Edit: Or by implementing DeBrujin indices. Thoughts from the reviewers on preference? We could also try to merge this sooner and then I can make a separate PR to implement DeBrujin (which, I may note, has most of the code relevant to the implementation in poly-adt-after, so it's a matter of making it work with the ADT code).

Edit2: Alternatively, we could require the same binding name in a type annotation and a type function. This seems to be the approach to most System F formalizations, though I'd say it's less free for the programmer. This amounts to requiring an equivalence between the typefun X and forall Y in the analytic rules for checking type functions.

Crazycolorz5 committed:
Add capture avoiding binding to recursive types.

Crazycolorz5 committed:
Add a comment to justify not needing capture avoidance in evaluator.

Crazycolorz5 commented:
![image](https://github.com/hazelgrove/hazel/assets/1497003/412ec86e-3b6b-4c79-b89a-e1aff36e31d4)

Use of type variable names can cause incorrect substitution when shadowing occurs. It's unclear to me how to sanely resolve this aside from alpha renaming on the spot if a typfun introduces shadowing, or to just use Debrujin indices.

Crazycolorz5 commented:
![image](https://github.com/hazelgrove/hazel/assets/1497003/c4299c75-58e2-4493-bde5-e796044ee98e)

Printing does not properly paranthesitize foralls. Also, printed foralls are "Forall" with capital F, but the syntax is with lowercase.

Also create an example page.

Crazycolorz5 committed:
Add parentheses to left type of functions when necessary. Lowercase forall and rec.

Crazycolorz5 committed:
Disallow shadowing with type variables.

Crazycolorz5 committed:
Add disallowing of shadowing for type variables.

Crazycolorz5 committed:
Use . (as opposed to ->) to delineate forall/rec binding from body.

Crazycolorz5 committed:
Make forall/rec bind tighter than sum to fix parsing bug.

Crazycolorz5 committed:
Change display information to "type binding" instead of alias or variable.

Crazycolorz5 committed:
Make shadowing typfun use the same case as a hole in tvar name.

Crazycolorz5 committed:
Add () around tuple output.

Crazycolorz5 committed:
Add Example scratchpad for polymorphism and recursive types.

Crazycolorz5 committed:
Correct minor error in is_tvar (should not affect anywhere where it's currently used)

Crazycolorz5 committed:
Preserve binding names from synthetic to give more intuitive cursor information.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

cyrus- commented:
What necessitated moving to `.` instead of `->` in forall types?

cyrus- commented:
> we could require the same binding name in a type annotation and a type function

I think this is the best solution here -- needing / exposing internally generated names is not ergonomic.

cyrus- commented:
It would be nice to allow lowercase type variable names with this change -- let's go ahead and change the regexp for type variable names. It would be good actually to only allow them in foralls (with implicit polymorphism, lowercase names will be implicitly quantified).

cyrus- commented:
Still need to add lang docs for the things in this PR.

Crazycolorz5 commented:
> What necessitated moving to . instead of -> in forall types?

It's not a necessary change, but I think it's useful for the reasons of:
- More closely matches how it's typically presented in theory. (In fact, I was motivated to change this initially because the type output printed them as such, and I didn't notice that it differed from how we input it until I was specifically trying to make them match)
- Disambiguates forall quantification from function types
- Makes a syntactic distinction between type functions (typfun, ->) and the type of type functions (forall, (.)). (I had actually found myself typing foralls where a typfun was expected. I suspect requiring a different separator for foralls will help with this, especially with those less familiar with operating on types).

> same binding name in a type annotation and a type function

> lowercase type variable names 

> lang docs

Sure, I can look at getting these done before Tuesday.

Crazycolorz5 committed:
Allow type names to be lowercase.

Crazycolorz5 committed:
Improve clarity of type outputs by fixing some parentheses.

Crazycolorz5 committed:
Add langdocs and examples for forall and rec.

Crazycolorz5 commented:
Re: same binding name in type annotation and type function.
Supporting this is more difficult than I initially thought. This is because we would like to require the strict equality in the analysis case. However, with the way Info is structured, we'd need to pass down information that it was in the analysis of a typfun (as opposed to the analysis for the argument of a function). Otherwise we end up excluding cases we don't want to, such as `let id : forall A . (A -> A) = in (fun x : forall X . (X -> X) -> x(id))` (which we don't want to exclude since this may be hidden behind several substitutions through which we may not be able to guarantee agreeing binding names). I've implemented this as-is in poly-adt-after3 if you're curious as to how that looks.

This is reflected in the formalism in Ap() and Cast() allowing the first type to be up to alpha. We don't have a good way to specifically allow this in Ana here, unless we add that to the mode, which seems a bit heavy for a (possibly?) niche issue that can be fixed with Debrujin indices.

For these reasons I'm going to leave this out of the current set of edits from the review.

Crazycolorz5 committed:
Add langdoc for typfun and typap. TODO more cases for typfun.

Crazycolorz5 commented on src/haz3lweb/view/LangDoc.re:
```diff
@@ -672,20 +672,78 @@ let get_doc =
-        );
+      | TypFun(_, _ /* tpat, body */) =>
+        basic_info(LangDocMessages.typ_function_group)
+      /* TODO: Cannot quite get the below to work. The EmptyHole case throws an unhandled exception Not_found.
```
Might need some help here from someone that understands the LangDoc infrastructure better than me?

Crazycolorz5 committed:
Keep original variable name in alpha rename.

Crazycolorz5 committed:
Fix error in previous commit.

Crazycolorz5 committed:
Return forall delimiter to -> from .

Crazycolorz5 committed:
Fix error.

cyrus- commented on src/haz3lcore/statics/Term.re:
```diff
@@ -128,6 +200,23 @@ module UTyp = {
+            {name, id: UTPat.rep_id(utpat), kind: Abstract},
+          );
+        Forall(name, to_typ(ctx, tbody));
+      | Forall(_, tbody) => to_typ(ctx, tbody)
```
* don't use underscore here in case of future expansion of UTPat
* if there is a UTPat hole here, then I think it should still elaborate to a forall -- we can either generate a fresh variable, or preferably add UTPat holes into the Typ representation as well. It should behave just like a variable that is never used.

Crazycolorz5 replied:
Do you just mean give the variable a name? I think the linter complains of it being an unused variable since we don't use it in the case where it's not a type variable name.

cyrus- commented on src/haz3lcore/statics/Term.re:
```diff
@@ -128,6 +200,23 @@ module UTyp = {
+            {name, id: UTPat.rep_id(utpat), kind: Abstract},
+          );
+        Rec(name, to_typ(ctx, tbody));
+      | Rec(_, tbody) => to_typ(ctx, tbody)
```
same here

cyrus- commented on src/haz3lcore/statics/Term.re:
```diff
@@ -548,19 +606,21 @@ module UExp = {
     | Let => "Let expression"
     | TyAlias => "Type Alias definition"
     | Ap => "Application"
+    | TypAp => "Type Application"
```
lowercase "a" in "application"

cyrus- commented on src/haz3lcore/statics/Term.re:
```diff
@@ -571,9 +631,13 @@ module UExp = {
     | UnOp(op) => show_unop(op)
     | Match => "Case expression";
 
+  // TODO (poly): May need to create a separate is_typfun function,
```
this sounds sensible to me

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -130,12 +133,43 @@ module rec Typ: {
+    | Prod(tys) => ListUtil.flat_map(free_vars(~bound), tys)
+    | Rec(x, ty) => free_vars(~bound=[x, ...bound], ty)
+    | Forall(x, ty) => free_vars(~bound=[x, ...bound], ty)
+    /* TODO: check that these are correct. */
```
lgtm

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -601,4 +648,33 @@ and Kind: {
   type t =
     | Singleton(Typ.t)
     | Abstract;
+  /* Is not needed as not using debrujin indices */
```
can delete if not used in this PR

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -175,40 +224,22 @@ module rec Typ: {
-      ConstructorMap.equal(Option.equal(eq), sm1, sm2)
+      ConstructorMap.equal(Option.equal(eq_internal(n)), sm1, sm2)
     | (Sum(_), _) => false
     | (Var(n1), Var(n2)) => n1 == n2
```
this does not normalize singletons, so that should be noted in a comment -- this is only type equivalence on already normalized types

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -244,18 +275,17 @@ module rec Typ: {
       !resolve && eq(ty_name, ty_join) ? Var(name) : ty_join;
     /* Note: Ordering of Unknown, Var, and Rec above is load-bearing! */
     | (Rec(x1, ty1), Rec(x2, ty2)) =>
-      /* TODO:
```
did this TODO get resolved? still looks like we're doing substitution with an open term, `Var(x1)`...

Crazycolorz5 replied:
I think this is resolved since subst() itself performs capture-avoiding substitution. I'll note that in a comment and replace this one.

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -365,6 +397,13 @@ module rec Typ: {
+  let matched_forall = (ctx, ty) =>
+    switch (weak_head_normalize(ctx, ty)) {
+    | Forall(t, ty) => (t, ty)
+    | Unknown(SynSwitch) => ("matched_forall", Unknown(SynSwitch))
```
this could be resolved if we added a TPat hole in the Typ representation -- don't need a fresh/dummy variable

cyrus- commented on src/haz3lcore/statics/TypBase.re:
```diff
@@ -442,6 +482,7 @@ and Ctx: {
   let lookup_var: (t, string) => option(var_entry);
   let lookup_ctr: (t, string) => option(var_entry);
   let is_alias: (t, TypVar.t) => bool;
+  let is_tvar: (t, TypVar.t) => bool;
```
should be called is_abstract not is_tvar

cyrus- commented on src/haz3lcore/statics/Kind.re:
```diff
@@ -1 +0,0 @@
-include TypBase.Kind;
```
don't need to delete this, useful to have as a convenience

cyrus- reviewed (changes requested):
partial review -- see comments

Crazycolorz5 committed:
Fix error in syntactix translation of types and address review comments.

Crazycolorz5 committed:
Exhaust case matching over using _ in Term

Crazycolorz5 committed:
Remove unused code.

cyrus- committed:
Merge branch 'dev' into poly-adt-after2

cyrus- commented:
@Crazycolorz5 when you get a chance, this needs some merge conflicts resolved.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

Crazycolorz5 commented:
<s>I addressed all the merge conflicts, but there seem to be a few subtle bugs with respect to argument casts on function applications after casts. See the following:
![image](https://github.com/hazelgrove/hazel/assets/1497003/cc79c4ec-e765-4807-8c28-41473a3512c9)

The reason this errors is because (fun x -> string_of_int(x+1))  is cast to `? -> ?`, so the literal `1` in the list is cast to `?`, but it then errors when it is plugged into `x+1`, as there is no cast added to turn it from `? => Int`. 

I'm not immediately sure how to fix this. Doing any of the following removes the error:
 - annotate x in the last line with `: ?`
 - fill in the type annotation for f as `X -> ?`
 - fill in the type annotation for f as `X -> Y` and annotate x as `: Int`
 
The following do not remove the error:
 - fill in the annotation for x as `: Int`
 - annotate x as `: ?` and f as `X -> ?`

The minimal example I can get to error from something like this is as follows:
![image](https://github.com/hazelgrove/hazel/assets/1497003/d32320be-cee2-4dfa-9e85-487b4af3a1e7)

Notably, this does not error on `dev`.

Given that Elaborator.re looks almost identical, I suspect it may be a bug in TypeBase.re? </s>


I fixed it in the commit just below.

Crazycolorz5 committed:
Fix error with missing casts from inappropriate synthesis joining.

cyrus- committed:
regenerate Init.ml

cyrus- reviewed (changes requested):
We're getting `Warning: Molds.get: can't decide if operand or operator: '.'` in the console on load, which doesn't happen on `dev`. Not sure what happened with dot, maybe from a previous iteration where we used dot as the forall separator?

Crazycolorz5 commented:
I've been looking into this warning for a few hours. I suspect it might have to do something with float tokenization in langdocs or init? I know for sure it's not caused by the forall stuff; I triple checked every where that that's parsed and it uses `->`. It could perhaps be a bug due to parser precedence as well? But I don't think I'm able to figure out what exactly is causing it (without better debugging tools for Reason). See the attached files for the console output when I print all tiles being re-molded (i.e. adding `print_string(Tile.show(t));` at line 129 in Segment.re, one for when I do this in `dev` and one here.
[out-dev.txt](https://github.com/hazelgrove/hazel/files/13847731/out-dev.txt)
[out-poly-adt-after2.txt](https://github.com/hazelgrove/hazel/files/13847732/out-poly-adt-after2.txt)

cyrus- commented:
As discussed, this needs to be brought up-to-date with dev, which has a new version of the LangDocs (ExplainThis) system that might fix the above issue. Marking as draft for now.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2.
Add langdocs for the new constructs.

Crazycolorz5 commented:
Updated to merge with dev. Note that #1215 breaks the current Polymorphism entry in the Documentation tab in Init. The same `Molds.get` warning from before is present; this seems to be benign and doesn't cause any errors.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

Crazycolorz5 commented:
![image](https://github.com/hazelgrove/hazel/assets/1497003/2c7250ba-ffe3-4a43-ba6b-f1d2e28c651c)
Was playing around with this example and this output is wrong. The variable is getting treated as a hole, which is correct, but it's not marked as such in the UI.


Edit: Seems like there's also an error with a forall type hole when rendering its text description
![image](https://github.com/hazelgrove/hazel/assets/1497003/02456342-0826-47b6-b8ef-71566c431cb5)

Crazycolorz5 committed:
Fix InvalidBoxedTypFun error on indeterminate argument in function position.

Crazycolorz5 committed:
Fixed error of type applications' type not getting proper Info.

Crazycolorz5 commented:
Fixed the errors noted in the previous comment.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

Crazycolorz5 committed:
Fix unresolved merge error (oops)

cyrus- commented:
Few small issues:

1. Type functions do not indent their bodies when you create a newline, the way regular functions do.

2. The ExplainThis description for type functions uses function syntax at the top, and let's use lowercase type variables in the examples.
<img width="471" alt="image" src="https://github.com/hazelgrove/hazel/assets/280638/65fcfaa2-1006-4421-8fd1-d222c88a0256">

3. The ExplainThis description for type application shouldn't use `_typ` for the type part, and the example is falling off the right side of the sidebar. Would be better to bind polymorphic identity to a variable and then instantiate it on the next line.

4. Need to add examples of type functions and type application to the Language Reference page in Documentation. (I can do this before PR merge if you'd like.)

cyrus- commented:
Also would be good to give polymorphic functions optional names in the internal language for display in the same way functions do, so it doesn't just always say anonymous typfun:
<img width="1043" alt="image" src="https://github.com/hazelgrove/hazel/assets/280638/5e558817-6a09-4b05-ae86-04a523180dcd">

(also one more merge conflict to resolve)

Crazycolorz5 committed:
Improve explainthis for type functions / applications.

Crazycolorz5 committed:
Add indent after typfun.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

Crazycolorz5 committed:
Add name to data record for dhexp typfuns.

Crazycolorz5 committed:
Fix a few remaining errors.

cyrus- commented:
The name still doesn't seem to show up in the result?
![image](https://github.com/hazelgrove/hazel/assets/280638/c3d51148-7e8c-435e-ae6e-16fe1ba55f1a)

Crazycolorz5 commented:
I think your screenshot is missing the result line, unless I'm misinterpreting where you mean for the name to show up?
This is what the version on build/poly-ady-after2 looks like for me:
![image](https://github.com/hazelgrove/hazel/assets/1497003/0d79a71f-276e-4805-acae-098fd947f89e)

Crazycolorz5 committed:
Add polymorphism to Documentation > Basic Reference.

Crazycolorz5 committed:
Differentiate type fun application and moving typap past cast.

cyrus- commented:
Marking as draft pending resolution of stepper issue we discussed today

Crazycolorz5 committed:
Rename a variable.

Crazycolorz5 committed:
Add TypFun case to unwrap.

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

cyrus- commented:
It would be nice if type application didn't erase the name of the function, e.g. if `<map>@<Int>` stepped to `<map@<Int>>` (essentially just updating the string field of the underlying function or type function on type app).

Crazycolorz5 committed:
Add type pretty printing and use it to augment typfun names when they are applied.

Crazycolorz5 commented:
![image](https://github.com/hazelgrove/hazel/assets/1497003/13996cb0-a0ce-464b-93cc-d84d65de36e5)
It has been done.

cyrus- committed:
add typfun expanding form

cyrus- committed:
remove dead code related to expanding forms

cyrus- commented:
looks like even non-recursive polymorphic functions are getting fixpoints placed around them:
![image](https://github.com/hazelgrove/hazel/assets/280638/a38d1410-c822-478f-ae5e-9ec6b8d0338e)

also the `Forall` there should not be capitalized and should not have the `.{tau}` notation

Crazycolorz5 commented:
> looks like even non-recursive polymorphic functions are getting fixpoints placed around them: ![image](https://private-user-images.githubusercontent.com/280638/321849855-a38d1410-c822-478f-ae5e-9ec6b8d0338e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMjU3NjQsIm5iZiI6MTcxMzMyNTQ2NCwicGF0aCI6Ii8yODA2MzgvMzIxODQ5ODU1LWEzOGQxNDEwLWM4MjItNDc4Zi1hZTVlLTllYzZiOGQwMzM4ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxN1QwMzQ0MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYmY3M2U0OTE4ZTEyM2E5Mzg3OTJmMzM2MDcyYTMyNGFjYTJjMjE1MDBhOWIzZDMzMDA3NjI3YjA5NmUwMmI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mBAPf5uG_lpQkJr4JS3jKzU6yznXS49rxQoRWEJoXuE)
> 
> also the `Forall` there should not be capitalized and should not have the `.{tau}` notation

![image](https://github.com/hazelgrove/hazel/assets/1497003/1f9de244-998d-4b01-8817-eaec26f6711b)
It appears that all functions that are type annotated have the fixed points -- this screenshot is from dev? As far as I can tell this is intended behavior,

I'll address the internal printing being Forall momentarily.

cyrus- commented:
Ah okay, I'll make a separate issue.

Crazycolorz5 committed:
Improve printing.

cyrus- commented:
@Crazycolorz5 looks nearly ready for merge -- just needs the merge conflict resolved!

Crazycolorz5 committed:
Merge branch 'dev' into poly-adt-after2

Crazycolorz5 commented:
Yep, just pushed that. Though I'm concerned about the potential for a large number of merge conflicts for others since we need to re-order the modules in `Term.re` since UTyp depends on UTPat in this branch. Nevertheless, this should be good, pending tests.

cyrus- commented:
Yeah no worries about merge conflicts with other branches, that's just part of the cost of having a lot of projects going on.
